### PR TITLE
refactor: make route() take a predicate for path_avoid

### PR DESCRIPTION
**Pull request by nornagon**
#### Summary
None

#### Purpose of change

Refactor map::route for a minor performance win and to support future
simplifications and bugfixes.

#### Describe the solution

**Before**
`map::route` takes a set of points to "preclose", i.e. never to path through.
Whenever calling `map::route`, callers assemble the full "preclosed" set
containing every point to avoid.

**After**
`map::route` takes a predicate that returns whether a point should be avoided.
This allows callers to avoid the overhead of assembling the full "preclosed"
set. In cases where paths are short, this may improve performance by not
considering points outside a very small radius. In cases where paths are long,
it should still improve performance in most cases as the predicate is called
only once per point that the pathfinding algorithm considers, which may be
smaller than the full set of points to avoid. Additionally, callers no longer
have to choose a radius in which to compute points to be avoided.

In some cases this might be slower than the previous approach, specifically
because the "avoid" calculation isn't cached, and may be called up to 8 times
per tile in pathological cases (one for each direction from which the pather
could enter the tile).

This change should have no observable effect on behavior. However, in future,
I'd like to refactor again to change the predicate into a function returning a
weight, which would allow expressing the concept "avoid this point if possible,
but if it's the only way to reach the destination, it's fine".

See #71532.

### CODE REVIEW

File: src/map.h [hunk 1]
```diff
@@ -729,10 +730,14 @@ class map
         // TODO: fix point types (remove the first overload)
         std::vector<tripoint> route( const tripoint &f, const tripoint &t,
                                      const pathfinding_settings &settings,
-        const std::unordered_set<tripoint> &pre_closed = {{ }} ) const;
+        std::function<bool( const tripoint & )> avoid = []( const tripoint & ) {
```
Comment (irwiss) [inline src/map.h:733]: If you pass the function as `const std::function<...> &avoid` instead you can avoid copies, also you can pass them directly without the lambda wrappers like in monmove.cpp line 1032, pathfinding.cpp 215 etc

File: src/map.h [hunk 2]
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#include "map_iterator.h"
```
[the remark this answers is not in the document]
Comment (nornagon) [reply src/map.h]: whoops, gonna blame VS Code's auto-include handling for this one 🙈

File: src/map.h (hunk 1 above)
Comment (nornagon) [reply src/map.h:733]: i think the copy is quite cheap here generally (should just be ~16 bytes most of the time so can happen on the stack even), but we can elide it so we might as well :)

**Comment by Maleclypse** [comment]
Conflict needs resolving

**Comment by nornagon** [comment]
test failure seems unrelated.

**Pull request merged**